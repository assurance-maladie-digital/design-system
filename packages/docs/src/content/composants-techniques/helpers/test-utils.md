---
title: testUtils
description: Le helper `testUtils` contient un ensemble de fonctions utilitaires pour les tests unitaires.
---

<doc-tabs>

<doc-tab-item label="Utilisation">

<doc-alert type="info">

Le helper `testUtils` exporte également toutes les fonctions du package [Vue Test Utils](https://vue-test-utils.vuejs.org/).

</doc-alert>

### applyGlobalConfig

La fonction `applyGlobalConfig` permet d’appliquer la configuration globale pour les tests, elle est utilisée dans le fichier `setup.ts`.

```ts
import { applyGlobalConfig } from '@cnamts/vue-dot/src/helpers/testUtils';

applyGlobalConfig();
```

### installGlobalPlugins

La fonction `installGlobalPlugins` permet d’installer les plugins [Vue Input Facade](https://ronaldjerez.github.io/vue-input-facade/), [Vue Meta](https://vue-meta.nuxtjs.org/), [Vuex](https://vuex.vuejs.org/), [Vuetify](https://v2.vuetifyjs.com/) et Vue Dot pour les tests.

```ts
import { createLocalVue, installGlobalPlugins } from '@cnamts/vue-dot/src/helpers/testUtils';

const localVue = createLocalVue();

installGlobalPlugins(localVue);
```

### installRouter

La fonction `installRouter` permet d’installer [Vue Router](https://v3.router.vuejs.org/fr/) pour les tests.

```ts
import { createLocalVue, installRouter } from '@cnamts/vue-dot/src/helpers/testUtils';

const localVue = createLocalVue();

installRouter(localVue);
```

### createRouter

La fonction `createRouter` permet de créer une instance de [Vue Router](https://v3.router.vuejs.org/fr/) pour les tests.

```ts
import { createRouter } from '@cnamts/vue-dot/src/helpers/testUtils';

const routes = [];

const router = createRouter({
	routes
});
```

### createStore

La fonction `createStore` permet de créer une instance de [Vuex](https://vuex.vuejs.org/) pour les tests.

```ts
import { createStore } from '@cnamts/vue-dot/src/helpers/testUtils';

const store = createStore({
	state: {}
});
```

### createVuetifyInstance

La fonction `createVuetifyInstance` permet de créer une instance de [Vuetify](https://v2.vuetifyjs.com/) pour les tests.

```ts
import { createVuetifyInstance } from '@cnamts/vue-dot/src/helpers/testUtils';

const vuetify = createVuetifyInstance();
```

### mockVFormRef

La fonction `mockVFormRef` permet de mocker la référence d’un composant [VForm](https://v2.vuetifyjs.com/en/components/forms/) pour les tests.

```ts
import { mockVFormRef } from '@cnamts/vue-dot/src/helpers/testUtils';

const mocks = {
	$refs: {
		form: mockVFormRef(true)
	}
};
```

### mockTranslations

La fonction `mockTranslations` permet de mocker les traductions pour les tests.

```ts
import { mockTranslations } from '@cnamts/vue-dot/src/helpers/testUtils';

const mocks = {
	...mockTranslations<string | string[]>({
		'path.to.title': 'Titre',
		'path.to.items': [
			'Élément 1',
			'Élément 2',
			'Élément 3'
		]
	})
};
```

</doc-tab-item>

<doc-tab-item label="API">
<doc-api name="helpers/test-utils"></doc-api>
</doc-tab-item>

</doc-tabs>
