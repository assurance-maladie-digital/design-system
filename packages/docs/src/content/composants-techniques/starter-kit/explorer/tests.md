Le dossier `tests` contient les tests d’interface et les tests unitaires globaux de l’application.

<doc-alert type="warning">
Il n’est pas recommandé de renommer ce dossier.
</doc-alert>
