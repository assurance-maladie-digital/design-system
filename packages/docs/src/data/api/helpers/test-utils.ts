import { Api } from '~/types';

export const api: Api = {
	applyGlobalConfig: {
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	installGlobalPlugins: {
		arguments: [
			{
				name: 'localVue',
				description: 'Local Vue instance',
				type: 'Vue',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	installRouter: {
		arguments: [
			{
				name: 'localVue',
				description: 'Local Vue instance',
				type: 'Vue',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	createRouter: {
		arguments: [
			{
				name: 'options',
				description: 'Router options',
				type: 'RouterOptions',
				required: false,
			},
		],
		returnValue: [
			{
				type: 'VueRouter',
			},
		],
	},
	createStore: {
		arguments: [
			{
				name: 'options',
				description: 'Store options',
				type: 'StoreOptions',
				required: false,
			},
		],
		returnValue: [
			{
				type: 'Store',
			},
		],
	},
	createVuetifyInstance: {
		arguments: [
			{
				name: 'preset',
				description: 'Vuetify preset',
				type: 'Partial<UserVuetifyPreset>',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'Vuetify',
			},
		],
	},
	mockVFormRef: {
		arguments: [
			{
				name: 'isValid',
				description: 'Is valid',
				type: 'boolean',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	mockTranslations: {
		arguments: [
			{
				name: 'translations',
				description: 'Translations',
				type: 'IndexedObject',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
};
