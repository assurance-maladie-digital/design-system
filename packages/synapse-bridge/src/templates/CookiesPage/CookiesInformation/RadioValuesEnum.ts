const RadioValuesEnum  = {
	ACCEPT:'accept',
	REJECT: 'reject',
}
export { RadioValuesEnum }
