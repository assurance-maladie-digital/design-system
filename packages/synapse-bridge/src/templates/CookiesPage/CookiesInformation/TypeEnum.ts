const TypeEnum  = {
	ESSENTIALS:'essentials',
	FUNCTIONAL: 'functional',
	ANALYTICS: 'analytics',
}

const TYPE_ENUM_VALUES = Object.values(TypeEnum)

export { TypeEnum, TYPE_ENUM_VALUES }
