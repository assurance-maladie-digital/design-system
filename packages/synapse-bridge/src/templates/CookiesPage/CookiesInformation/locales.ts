export const locales = {
	essentials: {
		title: 'Cookies requis',
		description: 'Ces cookies sont indispensables au fonctionnement du site et ne peuvent être désactivés.',
	},
	functional: {
		title: 'Cookies fonctionnels',
		description: 'Ces cookies sont utilisés pour capturer vos préférences sur notre site. Ils permettent de vous assurer un confort dans votre navigation.',
	},
	analytics: {
		title: 'Cookies d’analyse',
		description: 'Ces cookies sont dédiés à l’amélioration de notre site en analysant votre expérience d’utilisation. Ils nous sont très utiles pour vous proposer un site de bonne qualité.',
	},
	showInformation: 'Afficher les cookies',
	hideInformation: 'Masquer les cookies',
	reject: 'Refuser',
	accept: 'Accepter',
}
