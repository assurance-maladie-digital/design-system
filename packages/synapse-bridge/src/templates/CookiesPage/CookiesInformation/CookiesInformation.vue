<script lang="ts">
	import { defineComponent } from 'vue'
	import type { PropType } from 'vue'

	import { locales } from './locales'

	import { TypeEnum, TYPE_ENUM_VALUES } from './TypeEnum'
	import { RadioValuesEnum } from './RadioValuesEnum'

	import CookiesTable from '../CookiesTable'
	import { CookiesList } from '../CookiesTable/types'

	import { propValidator } from '../../../helpers/propValidator'

	import { required } from '../../../rules/required'

	import { mdiChevronDown, mdiChevronUp } from '@mdi/js'

	export default defineComponent({
		components: {
			CookiesTable,
		},
		emits: ['update:modelValue'],
		props: {
			type: {
				type: String as PropType<TypeEnum>,
				required: true,
				validator: (value: string) => propValidator('type', TYPE_ENUM_VALUES, value),
			},
			tableItems: {
				type: Array as PropType<CookiesList[]>,
				default: () => [],
			},
			modelValue: {
				type: Boolean as PropType<boolean | null>,
				default: null,
			},
		},
		data() {
			return {
				RadioValuesEnum: RadioValuesEnum,
				locales,
				rules: [required],
				open: false,
			}
		},
		computed: {
			parsedValue(): string | null {
				if (this.modelValue === null) {
					return null
				}

				return this.modelValue ? RadioValuesEnum.ACCEPT : RadioValuesEnum.REJECT
			},

			showRadios(): boolean {
				return this.type !== TypeEnum.ESSENTIALS
			},

			detailsLabel(): string {
				return this.open ? locales.hideInformation : locales.showInformation
			},

			arrowIcon(): string {
				return this.open ? mdiChevronUp : mdiChevronDown
			}
		},
		methods: {
			toggleDetails(event: Event): void {
				const { open } = event.target as HTMLDetailsElement
				this.open = open
			},

			emitChangeEvent(value: string): void {
				const parsedValue = Boolean(value === RadioValuesEnum.ACCEPT)
				this.$emit('update:modelValue', parsedValue)
			}
		}
	})
</script>

<template>
	<div class="vd-cookies-information">
		<h2 class="text-subtitle-1 font-weight-bold mb-2">
			{{ locales[type].title }}
		</h2>

		<p class="mb-4">
			{{ locales[type].description }}
		</p>

		<details @toggle="toggleDetails">
			<summary class="mb-1">
				{{ detailsLabel }}

				<VIcon class="mr-2">
					{{ arrowIcon }}
				</VIcon>
			</summary>

			<CookiesTable
				:items="tableItems"
				class="mb-2"
			/>
		</details>

		<VRadioGroup
			v-if="showRadios"
			:model-value="parsedValue"
			:rules="rules"
			data-test-id="radio-group"
			inline
			hide-details="auto"
			class="mt-0"
			@change="emitChangeEvent"
		>
			<VSpacer />

			<VRadio
				:label="locales.reject"
				:value="RadioValuesEnum.REJECT"
				color="primary"
			/>

			<VRadio
				:label="locales.accept"
				:value="RadioValuesEnum.ACCEPT"
				class="mr-0"
				color="primary"
			/>
		</VRadioGroup>
	</div>
</template>

<style lang="scss" scoped>
	details > summary {
		cursor: pointer;
		list-style: none;
		user-select: none;

		&::marker,
		&::-webkit-details-marker {
			display: none;
		}
	}

	.v-input--radio-group :deep(.v-messages) {
		text-align: right;
	}
</style>
