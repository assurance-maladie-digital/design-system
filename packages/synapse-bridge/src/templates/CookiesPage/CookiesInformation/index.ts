import CookiesInformation from './CookiesInformation.vue'

export default CookiesInformation
