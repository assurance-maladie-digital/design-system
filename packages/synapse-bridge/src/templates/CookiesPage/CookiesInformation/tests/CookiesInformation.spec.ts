import { describe, it, expect } from 'vitest'
import { mount, shallowMount } from '@vue/test-utils'
import { vuetify } from '@tests/unit/setup'

import CookiesInformation from '../'

import { cookies } from '../../tests/fixtures'
import { TypeEnum } from '../TypeEnum'
import { RadioValuesEnum } from '../RadioValuesEnum'

describe('CookiesInformation', () => {
	it('renders correctly', () => {
		const wrapper = shallowMount(CookiesInformation, {
			global: {
				plugins: [vuetify],
			},
			propsData: {
				tableItems: cookies,
				type: TypeEnum.FUNCTIONAL,
			},
		})

		expect(wrapper).toMatchSnapshot()
	})

	it('renders correctly without items', () => {
		const wrapper = shallowMount(CookiesInformation, {
			global: {
				plugins: [vuetify],
			},
			propsData: {
				type: TypeEnum.FUNCTIONAL,
			},
		})

		expect(wrapper).toMatchSnapshot()
	})

	it('renders correctly when details is toggled', async () => {
		const wrapper = mount(CookiesInformation, {
			global: {
				plugins: [vuetify],
			},
			propsData: {
				tableItems: cookies,
				type: TypeEnum.FUNCTIONAL,
			},
		})

		const details = wrapper.find('details')
		details.element.open = true

		await details.trigger('toggle')

		expect(wrapper).toMatchSnapshot()
	})

	it('parses the value when a radio button is selected or unselected', async () => {
		const wrapper = shallowMount(CookiesInformation, {
			global: {
				plugins: [vuetify],
			},
			propsData: {
				tableItems: cookies,
				type: TypeEnum.FUNCTIONAL,
				modelValue: true,
			},
		})

		expect(wrapper.vm.parsedValue).toBe(RadioValuesEnum.ACCEPT)

		await wrapper.setProps({
			modelValue: false,
		})

		expect(wrapper.vm.parsedValue).toBe(RadioValuesEnum.REJECT)
	})

	it('emits change event when a radio button is selected', async () => {
		const wrapper = mount(CookiesInformation, {
			global: {
				plugins: [vuetify],
			},
			propsData: {
				tableItems: cookies,
				type: TypeEnum.FUNCTIONAL,
			},
		})

		await wrapper.find('[data-test-id="radio-group"]').trigger('change', {
			value: RadioValuesEnum.ACCEPT,
		})

		await wrapper.vm.$nextTick()

		expect(wrapper.emitted()).toHaveProperty('change')
	})
})
