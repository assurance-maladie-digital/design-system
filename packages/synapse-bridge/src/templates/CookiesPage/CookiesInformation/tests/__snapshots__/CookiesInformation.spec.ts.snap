// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CookiesInformation > renders correctly 1`] = `
<div data-v-7eab3eb5="" class="vd-cookies-information">
  <h2 data-v-7eab3eb5="" class="text-subtitle-1 font-weight-bold mb-2">Cookies fonctionnels</h2>
  <p data-v-7eab3eb5="" class="mb-4">Ces cookies sont utilisés pour capturer vos préférences sur notre site. Ils permettent de vous assurer un confort dans votre navigation.</p>
  <details data-v-7eab3eb5="">
    <summary data-v-7eab3eb5="" class="mb-1">Afficher les cookies <v-icon-stub data-v-7eab3eb5="" class="mr-2" disabled="false" start="false" end="false" size="default" tag="i"></v-icon-stub>
    </summary>
    <cookies-table-stub data-v-7eab3eb5="" items="[object Object],[object Object]" class="mb-2"></cookies-table-stub>
  </details>
  <v-radio-group-stub data-v-7eab3eb5="" class="mt-0" height="auto" centeraffix="true" hidedetails="auto" hidespinbuttons="false" persistenthint="false" messages="" direction="horizontal" density="default" error="false" errormessages="" maxerrors="1" rules="(value) => {
    let valid;
    if (Array.isArray(value)) {
      valid = value.length !== 0;
    } else {
      valid = Boolean(typeof value === &quot;string&quot; ? value.trim() : value);
    }
    return valid || __vite_ssr_import_0__.ruleMessage(errorMessages, &quot;default&quot;);
  }" focused="false" inline="true" falseicon="$radioOff" trueicon="$radioOn" ripple="true" type="radio" valuecomparator="[Function]" data-test-id="radio-group"></v-radio-group-stub>
</div>
`;

exports[`CookiesInformation > renders correctly when details is toggled 1`] = `
<div data-v-7eab3eb5="" class="vd-cookies-information">
  <h2 data-v-7eab3eb5="" class="text-subtitle-1 font-weight-bold mb-2">Cookies fonctionnels</h2>
  <p data-v-7eab3eb5="" class="mb-4">Ces cookies sont utilisés pour capturer vos préférences sur notre site. Ils permettent de vous assurer un confort dans votre navigation.</p>
  <details data-v-7eab3eb5="" open="">
    <summary data-v-7eab3eb5="" class="mb-1">Masquer les cookies <i data-v-7eab3eb5="" class="M7.41,15.41L12,10.83L16.59,15.41L18,14L12,8L6,14L7.41,15.41Z mdi v-icon notranslate v-theme--light v-icon--size-default mr-2" aria-hidden="true"></i></summary>
    <div data-v-ad3c06ce="" data-v-7eab3eb5="" class="v-table v-theme--light v-table--density-default vd-cookie-table mb-2">
      <!---->
      <div class="v-table__wrapper">
        <table>
          <thead data-v-ad3c06ce="">
            <tr data-v-ad3c06ce="">
              <th data-v-ad3c06ce="" style="width: 180px;" class="text-left">Dénomination du cookie</th>
              <th data-v-ad3c06ce="" style="width: 328px;" class="text-left">Finalité</th>
              <th data-v-ad3c06ce="" style="width: 180px;" class="text-left">Durée de conservation</th>
            </tr>
          </thead>
          <tbody data-v-ad3c06ce="">
            <tr data-v-ad3c06ce="">
              <td data-v-ad3c06ce="">lorem-ipsum</td>
              <td data-v-ad3c06ce="">Lorem ipsum dolor sit amet, consectetur adipiscing elit.</td>
              <td data-v-ad3c06ce="">XX mois</td>
            </tr>
            <tr data-v-ad3c06ce="">
              <td data-v-ad3c06ce="">lorem-ipsum</td>
              <td data-v-ad3c06ce="">Lorem ipsum dolor sit amet, consectetur adipiscing elit.</td>
              <td data-v-ad3c06ce="">XX mois</td>
            </tr>
          </tbody>
        </table>
      </div>
      <!---->
    </div>
  </details>
  <div data-v-7eab3eb5="" class="v-input v-input--horizontal v-input--center-affix v-input--density-default v-locale--is-ltr v-radio-group mt-0" data-test-id="radio-group">
    <!---->
    <div class="v-input__control">
      <!---->
      <div class="v-selection-control-group v-selection-control-group--inline" role="radiogroup" aria-describedby="radio-group-0-messages">
        <div data-v-7eab3eb5="" class="v-spacer"></div>
        <div data-v-7eab3eb5="" class="v-selection-control v-selection-control--inline v-selection-control--density-default v-radio">
          <div class="v-selection-control__wrapper">
            <!---->
            <div class="v-selection-control__input"><i class="mdi-radiobox-blank mdi v-icon notranslate v-theme--light v-icon--size-default" aria-hidden="true"></i><input id="input-3" aria-disabled="false" type="radio" name="radio-group-0" value="reject"></div>
          </div><label class="v-label v-label--clickable" for="input-3">
            <!---->Refuser
          </label>
        </div>
        <div data-v-7eab3eb5="" class="v-selection-control v-selection-control--inline v-selection-control--density-default mr-0 v-radio mr-0">
          <div class="v-selection-control__wrapper">
            <!---->
            <div class="v-selection-control__input"><i class="mdi-radiobox-blank mdi v-icon notranslate v-theme--light v-icon--size-default" aria-hidden="true"></i><input id="input-4" aria-disabled="false" type="radio" name="radio-group-0" value="accept"></div>
          </div><label class="v-label v-label--clickable" for="input-4">
            <!---->Accepter
          </label>
        </div>
      </div>
    </div>
    <!---->
    <!---->
  </div>
</div>
`;

exports[`CookiesInformation > renders correctly without items 1`] = `
<div data-v-7eab3eb5="" class="vd-cookies-information">
  <h2 data-v-7eab3eb5="" class="text-subtitle-1 font-weight-bold mb-2">Cookies fonctionnels</h2>
  <p data-v-7eab3eb5="" class="mb-4">Ces cookies sont utilisés pour capturer vos préférences sur notre site. Ils permettent de vous assurer un confort dans votre navigation.</p>
  <details data-v-7eab3eb5="">
    <summary data-v-7eab3eb5="" class="mb-1">Afficher les cookies <v-icon-stub data-v-7eab3eb5="" class="mr-2" disabled="false" start="false" end="false" size="default" tag="i"></v-icon-stub>
    </summary>
    <cookies-table-stub data-v-7eab3eb5="" items="" class="mb-2"></cookies-table-stub>
  </details>
  <v-radio-group-stub data-v-7eab3eb5="" class="mt-0" height="auto" centeraffix="true" hidedetails="auto" hidespinbuttons="false" persistenthint="false" messages="" direction="horizontal" density="default" error="false" errormessages="" maxerrors="1" rules="(value) => {
    let valid;
    if (Array.isArray(value)) {
      valid = value.length !== 0;
    } else {
      valid = Boolean(typeof value === &quot;string&quot; ? value.trim() : value);
    }
    return valid || __vite_ssr_import_0__.ruleMessage(errorMessages, &quot;default&quot;);
  }" focused="false" inline="true" falseicon="$radioOff" trueicon="$radioOn" ripple="true" type="radio" valuecomparator="[Function]" data-test-id="radio-group"></v-radio-group-stub>
</div>
`;
